### PR TITLE
Cache ModelMetadata HashCode

### PR DESCRIPTION
**Pull request by ryanbrandenburg**
Add a cache to GetHashCode as per #5360.

This makes the assumption that ModelMetadata can't be changed once it's created.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelMetadata.cs [hunk 1]
```diff
@@ -405,9 +405,16 @@ public override bool Equals(object obj)
...
-            return Identity.GetHashCode();
+            if (_hashCode == null)
+            {
+                _hashCode = Identity.GetHashCode();
+            }
+
+            return _hashCode.Value;
         }
 
+        private int? _hashCode;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelMetadata.cs]: Well I'd just put in a comment that `Identity` is deeply immutable so it's safe to cache the hashcode.

**Comment by rynowak** [comment]
Perf results?

**Comment by ryanbrandenburg** [comment]
Perf results, as copied from #5360:

Scenario|Requests time|GetHashCode time|% time in GetHashCode
---------|--------|---------------|---
Without Cache|9417|550|5.84%
With Cache|6195|47|0.75%

**Comment by rynowak** [comment]
:shipit: - this seems like a good enough data set for such a straightforward change

**Comment by dougbu** [comment]
> assumption that `ModelMetadata` can't be changed once it's created

This isn't an assumption. `ModelMetadata` instances are immutable in practice. (Under the covers, property values are lazy-loaded but that's not important here.)

In any case, the `ModelMetadata` hash code is that of `ModelMetadataIdentity` and _that_ is 💯% immutable.

File: src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelMetadata.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelMetadata.cs]: Please move this to the top of the file.

**Pull request merged**